core/src/latest_unprocessed_votes.rs

```diff
@@ -0,0 +1,700 @@
+            | Ok(VoteInstruction::UpdateVoteStateSwitch(vote_state_update, _)) => {
+                let &pubkey = message
+                    .message
+                    .static_account_keys()
+                    .get(0)
+                    .ok_or(DeserializedPacketError::VoteTransactionError)?;
+                let slot = vote_state_update.last_voted_slot().unwrap_or(0);
+
+                Ok(Self {
+                    vote: Some(vote),
+                    slot,
+                    pubkey,
+                    vote_source,
+                    forwarded: false,
+                })
+            }
+            _ => Err(DeserializedPacketError::VoteTransactionError),
+        }
+    }
+
+    pub fn get_vote_packet(&self) -> Rc<ImmutableDeserializedPacket> {
+        self.vote.as_ref().unwrap().clone()
+    }
+
+    pub fn pubkey(&self) -> Pubkey {
+        self.pubkey
+    }
+
+    pub fn slot(&self) -> Slot {
+        self.slot
+    }
+
+    pub fn is_forwarded(&self) -> bool {
+        // By definition all gossip votes have been forwarded
+        self.forwarded || matches!(self.vote_source, VoteSource::Gossip)
+    }
+
+    pub fn is_processed(&self) -> bool {
+        self.vote.is_none()
+    }
+
+    pub fn clear(&mut self) {
+        self.vote = None;
+    }
+}
+
+pub fn deserialize_packets<'a>(
+    packet_batch: &'a PacketBatch,
+    packet_indexes: &'a [usize],
+    vote_source: VoteSource,
+) -> impl Iterator<Item = DeserializedVotePacket> + 'a {
+    packet_indexes.iter().filter_map(move |packet_index| {
+        DeserializedVotePacket::new(packet_batch[*packet_index].clone(), vote_source).ok()
+    })
+}
+
+pub(crate) fn weighted_random_order_by_stake<'a>(
+    bank: &Arc<Bank>,
+    pubkeys: impl Iterator<Item = &'a Pubkey>,
+) -> impl Iterator<Item = Pubkey> {
+    // Efraimidis and Spirakis algo for weighted random sample without replacement
+    let staked_nodes = bank.staked_nodes();
+    let mut pubkey_with_weight: Vec<(f64, Pubkey)> = pubkeys
+        .filter_map(|&pubkey| {
+            let stake = staked_nodes.get(&pubkey).copied().unwrap_or(0);
+            if stake == 0 {
+                None // Ignore votes from unstaked validators
+            } else {
+                Some((thread_rng().gen::<f64>().powf(1.0 / (stake as f64)), pubkey))
+            }
+        })
+        .collect::<Vec<_>>();
+    pubkey_with_weight.sort_by(|(w1, _), (w2, _)| w1.partial_cmp(w2).unwrap());
+    pubkey_with_weight.into_iter().map(|(_, pubkey)| pubkey)
+}
+
+pub struct VoteBatchInsertionMetrics {
+    num_dropped_gossip: usize,
+    num_dropped_tpu: usize,
+}
+
+#[derive(Debug, Default)]
+pub struct LatestUnprocessedVotes {
+    latest_votes_per_pubkey: RwLock<HashMap<Pubkey, RwLock<RefCell<DeserializedVotePacket>>>>,
+}
+
+unsafe impl Send for LatestUnprocessedVotes {}
+unsafe impl Sync for LatestUnprocessedVotes {}
+
+impl LatestUnprocessedVotes {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// Expensive
```

Comment:
A more detailed explanation would be helpful, maybe something like "Expensive because we have to iterate through and lock every item in the list to find the length"

Reply:
Could make this more efficient by incrementing a count in `update_latest_vote()` (only if update_latest_vote() actually moves from None -> Some) and removing the count in `clear()` (only if clear returns Some)

Reply:
Originally that's what we had with the separate atomic size variable. My understanding is that we didn't want to separately track size.